core/commands/repo.go

```diff
@@ -58,23 +65,41 @@ order to reclaim hard disk space.
 			return
 		}
 
-		gcOutChan, err := corerepo.GarbageCollectAsync(n, req.Context())
-		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
-			return
-		}
+		streamErrors, _, _ := res.Request().Option("stream-errors").Bool()
 
-		outChan := make(chan interface{})
+		gcOutChan := corerepo.GarbageCollectAsync(n, req.Context())
+
+		outChan := make(chan interface{}, cap(gcOutChan))
 		res.SetOutput((<-chan interface{})(outChan))
 
 		go func() {
 			defer close(outChan)
-			for k := range gcOutChan {
-				outChan <- k
+			unreportedError := false
+			var lastErr error
+			if streamErrors {
+				for res := range gcOutChan {
+					if unreportedError {
+						outChan <- &GcResult{Error: lastErr.Error()}
+						unreportedError = false
```

Comment:
Instead of trying to return the last error as the responses error, just return all errors through the outChan and if there were errors, have the response error be something like `"encountered errors during gc run"`